### PR TITLE
feat: Add Efficientdet Example

### PR DESCRIPTION
**Pull request by katport**
## Description

[DET-4500] Add Efficientdet Pytorch Example

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

## Title

feat: Add Efficientdet Example

[DET-4500]: https://determinedai.atlassian.net/browse/DET-4500

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
...
+| Computer Vision | <p>[cifar10_pytorch](computer_vision/cifar10_pytorch)<br>[fasterrcnn_coco_pytorch](computer_vision/fasterrcnn_coco_pytorch)<br>[mmdetection_pytorch](computer_vision/mmdetection_pytorch)<br>[detr_coco_pytorch](computer_vision/detr_coco_pytorch)<br>[mnist_estimator](computer_vision/mnist_estimator)<br>[mnist_tf_layers](computer_vision/mnist_tf_layers)<br>[cifar10_tf_keras](computer_vision/cifar10_tf_keras)<br>[iris_tf_keras](computer_vision/iris_tf_keras)<br>[unets_tf_keras](computer_vision/unets_tf_keras)<br>[efficientdet_pytorch](computer_vision/efficientdet_pytorch)</p> | <p>CIFAR-10<br>Penn-Fudan Dataset<br>COCO<br>MNIST<br>MNIST<br>CIFAR-10<br>Iris Dataset<br>Oxford-IIIT Pet Dataset</p> | <p>PyTorch<br>PyTorch<br>PyTorch<br>TensorFlow (Estimator API)<br>TensorFlow (tf.layers via Estimator API)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)</p> |
```
Comment (naren-determined) [inline examples/README.md:5]: You also have to add the framework and dataset here. Update to this:

`| Computer Vision | <p>[cifar10_pytorch](computer_vision/cifar10_pytorch)<br>[fasterrcnn_coco_pytorch](computer_vision/fasterrcnn_coco_pytorch)<br>[mmdetection_pytorch](computer_vision/mmdetection_pytorch)<br>[detr_coco_pytorch](computer_vision/detr_coco_pytorch)<br>[mnist_estimator](computer_vision/mnist_estimator)<br>[mnist_tf_layers](computer_vision/mnist_tf_layers)<br>[cifar10_tf_keras](computer_vision/cifar10_tf_keras)<br>[iris_tf_keras](computer_vision/iris_tf_keras)<br>[unets_tf_keras](computer_vision/unets_tf_keras)<br>[efficientdet_pytorch](computer_vision/efficientdet_pytorch)</p> | <p>CIFAR-10<br>Penn-Fudan Dataset<br>COCO<br>MNIST<br>MNIST<br>CIFAR-10<br>Iris Dataset<br>Oxford-IIIT Pet Dataset<br>COCO</p> | <p>PyTorch<br>PyTorch<br>PyTorch<br>TensorFlow (Estimator API)<br>TensorFlow (tf.layers via Estimator API)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)<br>PyTorch</p> |`

File: examples/computer_vision/efficientdet_pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+### Configuration Files
+* **const.yaml**: Train the model with constant hyperparameter values.
+* **distributed.yaml**: Same as `const.yaml`, but trains the model with multiple GPUs (distributed training).
+* **adaptive.yaml**: Runs an adaptive search with the configured restraints.
+
+
+## Data
+The coco dataset is the primary dataset for this port. Information on how to download the dataset can be found [here](https://github.com/rwightman/efficientdet-pytorch/tree/678bae1597eb083e05b033ee3eb585877282279a#coco). Once downloaded, copy the data on all the Determined agents. Then configure the bind mount in the configuration files to the appropriate location.
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/README.md:19]: Can you link the docs to bind_mount here?

File: examples/computer_vision/efficientdet_pytorch/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+## To Run
+If you have not yet installed Determined, installation instructions can be found
+under `docs/install-admin.html` or at https://docs.determined.ai/latest/index.html
+
+Run the following command: `det -m <master host:port> experiment create -f 
+const.yaml .`. The other configurations can be run by specifying the appropriate 
+configuration file in place of `const.yaml`.
+
+## Results
+The default configurations are based on [rwightman's reported results](https://github.com/rwightman/efficientdet-pytorch/tree/678bae1597eb083e05b033ee3eb585877282279a#efficientdet-d0). After 6 training days, the model `efficientdet_d0`'s mAP should be about .33. When running with `sync_batch_norm=True`, setting `tensor_fusion_threshold: 1` and `tensor_fusion_cycle_time: 2` will increase training performance time.
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/README.md:30]: Grammar Nitpick: Remove the `When` from the last sentence so it makes sense grammatically

File: examples/computer_vision/efficientdet_pytorch/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Note: The purpose of these graphs is to show the `efficientdet_d0` model running in Determined 
+for a set number of epochs, demonstrating the acceleration of model training time 
+achieved with Determined’s distributed training.
+
+![Single GPU vs. Distributed Training with Determined AI](.png)
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/README.md:36]: Assuming this .png is empty b/c the experiment is still in progress.

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+searcher:
+  name: adaptive_asha
+  metric: val_loss
+  max_length:
+      epochs: 300
+  smaller_is_better: False
+  max_trials: 1000
+  max_concurrent_trials: 4
+resources:
+  slots_per_trial: 2
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/adaptive.yaml:94]: Is there a reason this is just set to 2 and not something higher given most AWS Determined default agents have more resources?

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+reproducibility:
+  experiment_seed: 42
+optimizations:
+   tensor_fusion_threshold: 1
+   tensor_fusion_cycle_time: 2
+entrypoint: model_def:EffDetTrial
+max_restarts: 0
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/adaptive.yaml:101]: Remove `max_restarts` when ready to merge.

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+bind_mounts:
+  - host_path: /home/katie/projects/data/coco
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/adaptive.yaml:103]: Can you put a comment here basically asking the user to replace this path with their path to COCO and perhaps mention that in the README somehow or something?

File: examples/computer_vision/efficientdet_pytorch/const.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  metric: val_loss
+  max_length:
+    # batches: 2
+    epochs: 300
+reproducibility:
+  experiment_seed: 42
+entrypoint: model_def:EffDetTrial
+min_validation_period:
+  epochs: 1
+max_restarts: 0
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/const.yaml:78]: Remove max_restarts when ready to merge.

File: examples/computer_vision/efficientdet_pytorch/const.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+bind_mounts:
+  - host_path: /home/katie/projects/data/coco
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/const.yaml:80]: Ditto on the bind_mounts.

File: examples/computer_vision/efficientdet_pytorch/distributed.yaml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      epochs: 300
+  smaller_is_better: true
+resources:
+  slots_per_trial: 4
+reproducibility:
+  experiment_seed: 42
+min_validation_period:
+  epochs: 1
+entrypoint: model_def:EffDetTrial
+max_restarts: 0
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/distributed.yaml:83]: Ditto max_restarts.

File: examples/computer_vision/efficientdet_pytorch/distributed.yaml [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+bind_mounts:
+  - host_path: /home/katie/projects/data/coco
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/distributed.yaml:85]: Ditto bind_mounts

File: examples/computer_vision/efficientdet_pytorch/startup-hook.sh [hunk 12]
```diff
@@ -0,0 +1,16 @@
+cp -pv /run/determined/workdir/patch/__init__.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/pytorch/__init__.py
+cp -pv /run/determined/workdir/patch/_experimental.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/pytorch/_experimental.py
+
+cp -pv /run/determined/workdir/patch/_pytorch_trial.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/pytorch/_pytorch_trial.py
+cp -pv /run/determined/workdir/patch/_trial_controller.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/_trial_controller.py
+
+cp -pv /run/determined/workdir/patch/pytorch_context.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/pytorch/_pytorch_context.py
+cp -pv /run/determined/workdir/patch/reducer.py /run/determined/pythonuserbase/lib/python3.6/site-packages/determined/pytorch/_reducer.py
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/startup-hook.sh:8]: Assuming this is some hot patching you did that isn't necessary once the experiment is cleaned up and ready to merge?

File: examples/computer_vision/efficientdet_pytorch/efficientdet_files/evaluator.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
+'''
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/efficientdet_files/evaluator.py:1]: For any files in `efficientdet_files` directory you modified from the OG implementation, can you just make a quick comment at the top as to what the modifications were and why they were necessary?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+MAX_NUM_INSTANCES = 100
+
+
+class EffDetTrial(PyTorchTrial):
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        self.context = context
+        self.hparam = self.context.get_hparam
+        self.args = DotDict(self.context.get_hparams())
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:41]: You define class `DotDict` at the bottom of the model_def.py. Could you maybe move it to the top that way it's just easy to recall this usage here?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+            bench_task='train',
+            num_classes=self.args.num_classes,
+            pretrained=self.args.pretrained,
+            pretrained_backbone=self.args.pretrained_backbone,
+            redundant_bias=self.args.redundant_bias,
+            label_smoothing=self.args.smoothing,
+            new_focal=self.args.new_focal,
+            jit_loss=self.args.jit_loss,
+            bench_labeler=self.args.bench_labeler,
+        checkpoint_path=self.args.initial_checkpoint,
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:65]: Indent

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+            fill_color='mean',
+            mean=IMAGENET_DEFAULT_MEAN,
+            std=IMAGENET_DEFAULT_STD,
+            num_workers=4,
+            distributed=False,
+            pin_mem=False,
+            anchor_labeler=None,
+        ):
+
+            print ('input_size', input_size)
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:154]: Assuming this is for easy debugging -- eventually can we remove this?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+                img_size = input_size
+
+            if is_training:
+                transform = transforms_coco_train(
+                    img_size,
+                    interpolation=interpolation,
+                    use_prefetcher=use_prefetcher,
+                    fill_color=fill_color,
+                    mean=mean,
+                    std=std)
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:182]: This is super nitpicky and totally a style thing, but can we move the last parentheses around a function call to the next line if each argument is a separate line? See the definition of `loader` below.

**Review by naren-determined — CHANGES_REQUESTED** [review]
Looks good overall, will do another pass once some of the logging and what I assume to be debugging logic is cleaned a bit :)

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+            std=self.input_config['std'],
+            num_workers=self.args.workers,
+            distributed=self.args.distributed,
+            pin_mem=self.args.pin_mem,
+            anchor_labeler=self.labeler,
+        )
+        
+        self.evaluator = self.create_evaluator(self.args.dataset, self.loader_eval.dataset, pred_yxyx=False)
+
+        self.eval_counter = 0
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:277]: Does't seem like this is being used anywhere?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+        self.val_reducer.update(vals)
+        self.evaluator.reset()
+
+        return {'val_loss': reduced_loss}   
+
+    def create_evaluator(self,name, dataset, pred_yxyx=False):
+        if 'coco' in name:
+            return CocoEvaluator(dataset, pred_yxyx=pred_yxyx)
+            
+    def validation_reducer(self, values):
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:356]: Can we use custom reduce metrics here or should the implementation be left as this?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+            mod.running_mean = module.running_mean
+            mod.running_var = module.running_var
+            if module.affine:
+                mod.weight.data = module.weight.data.clone().detach()
+                mod.bias.data = module.bias.data.clone().detach()
+        for name, child in module.named_children():
+            mod.add_module(name, self.convert_syncbn_model(child,
+                                                    process_group=process_group,
+                                                    channel_last=channel_last))
+        # TODO(jie) should I delete model explicitly?
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:392]: Was this from the original code?

**Review by naren-determined — COMMENTED** [review]
Can you also add the nightly tests + docs update per https://determinedai.atlassian.net/wiki/spaces/ENGINEERIN/pages/479330312/Adding+an+Example+to+Determined?

File: examples/computer_vision/efficientdet_pytorch/model_def.py (hunk 20 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/model_def.py:392]: yes

File: examples/computer_vision/efficientdet_pytorch/model_def.py (hunk 19 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/model_def.py:356]: Offline disc: this is using the simple API

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+  pin_mem: False
+  no_prefetcher: False
+  bench_labeler: False
+  eval_metric: map
+  tta: 0
+  epochs: 300 # used for LR calculations
+records_per_epoch: 118287
+searcher:
+  name: adaptive_asha
+  mode: aggressive
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/adaptive.yaml:87]: Might be useful providing a comment here if this is something the user may want to toggle.

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+  min_loss_scale:
+    type: double
+    minval: 8
+    maxval: 64
+  sync_bn: 
+    type: categorical
+    vals: [True, False]
+  model_ema: True
+  amp: True
+  root: '/mnt/data'
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/adaptive.yaml:13]: I don't see this being used in any non-configuration files. Ditto in const.yaml and distributed.yaml.

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+
+from timm.optim import create_optimizer
+from timm.scheduler import create_scheduler
+
+from determined.pytorch import DataLoader, PyTorchTrial, PyTorchTrialContext, LRScheduler
+from determined.experimental import Determined
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+MAX_NUM_INSTANCES = 100
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:35]: Why is this necessary?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 24]
```diff
@@ -0,0 +1,394 @@
...
+            std=self.input_config['std'],
+            num_workers=self.args.workers,
+            distributed=self.args.distributed,
+            pin_mem=self.args.pin_mem,
+            anchor_labeler=self.labeler,
+        )
+        if self.model_config.num_classes < loader_train.dataset.parser.max_label:
+            logging.error(
+                f'Model {self.model_config.num_classes} has fewer classes than dataset {loader_train.dataset.parser.max_label}.')
+            exit(1)
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:246]: Should this be sys.exit()?

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 25]
```diff
@@ -0,0 +1,394 @@
...
+            mean=self.input_config['mean'],
+            std=self.input_config['std'],
+            num_workers=self.args.workers,
+            distributed=self.args.distributed,
+            pin_mem=self.args.pin_mem,
+            anchor_labeler=self.labeler,
+        )
+        if self.model_config.num_classes < loader_train.dataset.parser.max_label:
+            logging.error(
+                f'Model {self.model_config.num_classes} has fewer classes than dataset {loader_train.dataset.parser.max_label}.')
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:245]: From a style perspective, might be good to have the last parenthesis on a different line. Ditto below.

File: examples/computer_vision/efficientdet_pytorch/model_def.py [hunk 26]
```diff
@@ -0,0 +1,394 @@
...
+            std=self.input_config['std'],
+            num_workers=self.args.workers,
+            distributed=self.args.distributed,
+            pin_mem=self.args.pin_mem,
+            anchor_labeler=self.labeler,
+        )
+        
+        self.evaluator = self.create_evaluator(self.args.dataset, self.loader_eval.dataset, pred_yxyx=False)
+
+        self.eval_counter = 0
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/model_def.py:277]: This is still here, but the comment got resolved.

**Review by naren-determined — COMMENTED** [review]
Can you also add testing logic here if dataloading doesn't take too long? No need for convergence.

File: examples/computer_vision/efficientdet_pytorch/adaptive.yaml (hunk 21 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/adaptive.yaml:87]: done

File: examples/computer_vision/efficientdet_pytorch/model_def.py (hunk 23 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/model_def.py:35]: removed

File: examples/computer_vision/efficientdet_pytorch/model_def.py (hunk 26 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/model_def.py:277]: github did github things on merged, should be removed now.

File: examples/computer_vision/efficientdet_pytorch/model_def.py (hunk 25 above)
Comment (katport) [reply examples/computer_vision/efficientdet_pytorch/model_def.py:245]: done

File: examples/README.md [hunk 27]
```diff
@@ -11,4 +11,4 @@
 | Decision Trees  | [gbt_titanic_estimator](decision_trees/gbt_titanic_estimator) | Titanic | TensorFlow (Estimator API) |
 | Data Layer | <p>[data_layer_mnist_estimator](features/data_layer_mnist_estimator)<br>[data_layer_mnist_tf_keras](features/data_layer_mnist_tf_keras)</p> | <p>MNIST<br>MNIST</p> | <p>TensorFlow (Estimator API)<br>TensorFlow (tf.keras)</p> |
 | Custom Reducers | <p>[custom_reducers_mnist_pytorch](features/custom_reducers_mnist_pytorch)</p> | <p>MNIST</p> | <p>PyTorch</p> |
-| HP Search Constraints | <p>[hp_constraints_mnist_pytorch](features/hp_constraints_mnist_pytorch)</p> | <p>MNIST</p> | <p>PyTorch</p> |
+| HP Search Constraints | <p>[hp_constraints_mnist_pytorch](features/hp_constraints_mnist_pytorch)</p> | <p>MNIST</p> | <p>PyTorch</p> |
```
Comment (naren-determined) [inline examples/README.md:14]: Looks like nothing changed here? Maybe I'm missing something.

File: examples/computer_vision/efficientdet_pytorch/README.md [hunk 28]
```diff
@@ -0,0 +1,37 @@
+# EfficientDet Example
+
+This example is an EfficientDet port using Determined's PytorchTrial API. The original example can be found on
```
Comment (naren-determined) [inline examples/computer_vision/efficientdet_pytorch/README.md:3]: Nit: PyTorchTrial

**Review by naren-determined — COMMENTED** [review]
Can you add test logic to test_capability.py?

**Review by naren-determined — APPROVED** [review]
